teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md

```diff
@@ -0,0 +1,478 @@
+        - name: IMAGE_DIGEST
+          description: digest of image
+    ```
+    A Task example `with` structured results and params support will be like:
+    ```
+    apiVersion: tekton.dev/v1beta1
+    kind: Task
+    spec:
+      params:
+        - name: CHAINS-VCS
+          type: object
+          description: |
+            The source distribution
+            * uri: resource uri of the git commit
+            * digest: revision digest
+            * alg: digest algorithm: sha256 etc.
+            * src_type: source VCS type: git, hg, etc.
+          properties:
+          src_type:
+            type: string
+          uri:
+            type: string
+          alg:
+            type: string
+          digest:
+            type: string
+      ...
+      results:
+        - name: CHAINS-OCI_IMAGE
+          type: object
+          description: |
+            The source distribution
+            * name: name of the built image
+            * digest: image digest
+            * alg: digest algorithm: sha256 etc.
+          properties:
+          name:
+            type: string
+          digest:
+            type: string
+          alg:
+            type: string
+    ```
+    And its corresponding TaskRun example will be like:
+    ```
+    TaskRun:
+      ...
+      params:
+        - name: CHAINS-VCS
+          value:
+            src_type: git
+            uri: git+https://github.com/foo/bar.git
+            alg: sha256
+            digest: abc
+      ...
+      results:
+        - name: CHAINS-OCI_IMAGE
+          value:
+            name: gcr.io/somerepo/someimage
+            digest: abc
+            alg: sha256
+    ```
+
+2. Use Perforce and images as sources, and generates in-toto provenance for a TaskRun that builds a Python wheel.
+
+    TODO: Example to be added 
+
+### Requirements
+
+- Able to structurally retrieve provenance info from TaskRun and produces currently supported provenance formats once a TaskRun is finished
+- Based on SLSA L2 requirements, the provenance must provide the location of source code in version control and the provenance MUST identify the output artifact via at least one cryptographic hash.
+- If a TaskRun is within a PipelineRun, the TaskRun’s provenance can be produced between the time that when one TaskRun is finished
+
+
+## Proposal
+
+Currently we are using type-hinting to retrieve needed provenance info from TaskRun’s params and results. The proposal aims to leverage the new features of structured params and results described in [TEP 0075](https://github.com/tektoncd/community/blob/main/teps/0075-object-param-and-result-types.md) and [TEP 0076](https://github.com/tektoncd/community/blob/main/teps/0076-array-result-types.md) and support better structured provenance retrieval for inputs and artifacts to make the current type-hinting easier. This proposal is focused on the use cases for generation in-toto attestations, as other types of attestations don’t need the TaskRun params and results.
```

Comment:
Can you explain a bit more about ideas of such tool? Do you mean we provide an image for users or a piece of module in Chains?

Reply:
In my mind tool == any task (i.e. a git task). The Tasks would need to be responsible for outputting the relevant provenance for their given artifact type.

Chains wouldn't need to be aware of any particular tasks/images - it would just consume the provenance output for each task (similar to the result based mechanism it uses today).

Reply:
To reflect what we discussed offline, I updated the doc in proposal section. I added flexibility that the provenance can come from results and other new fields that's only for trusted tasks

Reply:
I like the idea of having something that generates the structure response because it allows users to extend support for a particular VCS without having to make any changes to Chains. It's just another Task that needs to be written.

With [TEP-84](https://github.com/tektoncd/community/blob/main/teps/0084-endtoend-provenance-collection.md) in place, post-build policies can be put in place to ensure that the Task that produced the VCS information is trusted.